python-package/lightgbm/basic.py

```diff
@@ -237,44 +237,44 @@ def _data_to_2d_numpy(data: Any, dtype: type = np.float32, name: str = 'list') -
                     "It should be list of lists, numpy 2-D array or pandas DataFrame")
 
 
-def cfloat32_array_to_numpy(cptr, length):
+def cfloat32_array_to_numpy(cptr: ctypes.POINTER, length: int) -> np.ndarray:
     """Convert a ctypes float pointer array to a numpy array."""
     if isinstance(cptr, ctypes.POINTER(ctypes.c_float)):
         return np.ctypeslib.as_array(cptr, shape=(length,)).copy()
     else:
         raise RuntimeError('Expected float pointer')
 
 
-def cfloat64_array_to_numpy(cptr, length):
+def cfloat64_array_to_numpy(cptr: ctypes.POINTER, length: int) -> np.ndarray:
     """Convert a ctypes double pointer array to a numpy array."""
     if isinstance(cptr, ctypes.POINTER(ctypes.c_double)):
         return np.ctypeslib.as_array(cptr, shape=(length,)).copy()
     else:
         raise RuntimeError('Expected double pointer')
 
 
-def cint32_array_to_numpy(cptr, length):
+def cint32_array_to_numpy(cptr: ctypes.POINTER, length: int) -> np.ndarray:
     """Convert a ctypes int pointer array to a numpy array."""
     if isinstance(cptr, ctypes.POINTER(ctypes.c_int32)):
         return np.ctypeslib.as_array(cptr, shape=(length,)).copy()
     else:
         raise RuntimeError('Expected int32 pointer')
 
 
-def cint64_array_to_numpy(cptr, length):
+def cint64_array_to_numpy(cptr: ctypes.POINTER, length: int) -> np.ndarray:
     """Convert a ctypes int pointer array to a numpy array."""
     if isinstance(cptr, ctypes.POINTER(ctypes.c_int64)):
         return np.ctypeslib.as_array(cptr, shape=(length,)).copy()
     else:
         raise RuntimeError('Expected int64 pointer')
 
 
-def c_str(string):
+def c_str(string: str) -> ctypes.c_char_p:
     """Convert a Python string to C string."""
     return ctypes.c_char_p(string.encode('utf-8'))
 
 
-def c_array(ctype, values):
+def c_array(ctype: type, values: List[ctypes.c_char_p]) -> ctypes.Array:
```

Comment:
I believe this is universal function, not just for strings.
For example,
https://github.com/dmlc/xgboost/blob/ff1c559084cf2b3732b8406109f2b39e9b665528/python-package/xgboost/core.py#L371-L377
https://github.com/dmlc/xgboost/blob/ff1c559084cf2b3732b8406109f2b39e9b665528/python-package/xgboost/core.py#L1427
https://github.com/dmlc/xgboost/blob/ff1c559084cf2b3732b8406109f2b39e9b665528/python-package/xgboost/core.py#L1552
```suggestion
def c_array(ctype: type, values: List[Any]) -> ctypes.Array:
```

Reply:
hmmm ok. I recommended `ctypes.c_char_p` because the only place that it is used in `lightgbm` is with a list of `ctypes.c_char_p`, like I mentioned in https://github.com/microsoft/LightGBM/pull/5362#discussion_r918444042.

I think that most accurately it is that `"`ctype` needs to be one of the pointer types exported by `ctypes`, and `values` should be a list of corresponding instances that those pointers can refer to`.

But I understand that since there is no base class for those types in the `ctypes` namespace (just the internal `_ctypes.PyCSimpleType`), `Any` might be the best we can do.

```python
from ctypes import c_char_p
type(c_char_p)
# <class '_ctypes.PyCSimpleType'>

type(c_char_p).__mro__
# (<class '_ctypes.PyCSimpleType'>, <class 'type'>, <class 'object'>)
```

I'm ok with this change. @makquel will you please accept this suggestion?

Reply:
I just applied this suggestion, to hopefully move this PR forward.